app/clients.go

```diff
@@ -0,0 +1,128 @@
+package app
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/OAuth2withJWT/identity-provider/app/validation"
+)
+
+type ClientService struct {
+	repository ClientRepository
+}
+
+type Client struct {
+	Id          string
+	Name        string
+	Secret      string
+	Scope       string
+	RedirectURI string
+	CreatedBy   int
+}
+
+type CreateClientRequest struct {
+	Name        string
+	Scope       string
+	RedirectURI string
+	CreatedBy   int
+}
+
+type ClientCredentials struct {
+	Id     string
+	Secret string
+}
+
+func (req *CreateClientRequest) validateClientRegistrationFields(s *ClientService) error {
+	v := validation.New()
+	v.IsEmpty("Client Name", req.Name)
+	v.IsEmpty("Scope", req.Scope)
+	v.IsEmpty("Redirect URI", req.RedirectURI)
+	v.IsValidURI("Redirect URI", req.RedirectURI)
+	if s.hasClientWithName(req.Name) {
+		v.AddError("Client Name", fmt.Errorf("Client with that name already exists"))
+	}
+
+	if s.hasClientWithRedirectURI(req.RedirectURI) {
+		v.AddError("Redirect URI", fmt.Errorf("Client with that redirect URI already exists"))
+	}
+
+	return v.Validate()
+}
+
+func (s *ClientService) hasClientWithName(name string) bool {
+	client, _ := s.repository.GetClientByName(name)
+	return client != (Client{})
+}
+
+func (s *ClientService) hasClientWithRedirectURI(redirectURI string) bool {
+	client, _ := s.repository.GetClientByRedirectURI(redirectURI)
+	return client != (Client{})
+}
+
+func (s *ClientService) Create(req CreateClientRequest) (*Client, error) {
+
+	err := req.validateClientRegistrationFields(s)
+	if err != nil {
+		return nil, err
+	}
+
+	clientID, err := s.generateClientCredentials(16)
+	if err != nil {
+		return nil, err
+	}
+
+	clientSecret, err := s.generateClientCredentials(32)
+	if err != nil {
+		return nil, err
+	}
+
+	credentials := ClientCredentials{clientID, clientSecret}
+
+	newClient, err := s.repository.Create(req, credentials)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return newClient, nil
+}
+
+func NewClientService(cr ClientRepository) *ClientService {
+	return &ClientService{
+		repository: cr,
+	}
+}
+
+type ClientRepository interface {
+	Create(req CreateClientRequest, credentials ClientCredentials) (*Client, error)
+	GetClientByName(name string) (Client, error)
+	GetClientByRedirectURI(redirectURI string) (Client, error)
+}
+
+func (s *ClientService) GetClientByName(name string) (Client, error) {
+	client, err := s.repository.GetClientByName(name)
+	if err != nil {
+		return Client{}, err
+	}
+
+	return client, nil
+}
+
+func (s *ClientService) GetClientByRedirectURI(redirectURI string) (Client, error) {
+	client, err := s.repository.GetClientByName(redirectURI)
+	if err != nil {
+		return Client{}, err
+	}
+
+	return client, nil
+}
+
+func (s *ClientService) generateClientCredentials(length int) (string, error) {
```

Comment:
Maybe call the function `generateRandomBytes` because that's what we're doing. You're not generating credentials, so the naming is misleading.